File: WebAPI/routers/watchlist.py
# -*- coding: utf-8 -*-
"""
Watchlist 路由 — 自选分组管理 API。
"""

from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from ..watchlist_store import (
    add_stock,
    create_folder,
    delete_folder,
    get_folder_stocks,
    list_folders,
    move_stock,
    remove_stock,
    rename_folder,
)

router = APIRouter(prefix="/api/watchlist", tags=["watchlist"])


@router.get("/folders")
async def watchlist_folders():
    """自选分组列表。"""
    return list_folders()


@router.post("/folders")
async def watchlist_create_folder(body: dict):
    """新建分组。"""
    name = body.get("name", "")
    if not name:
        raise HTTPException(status_code=422, detail="name is required")
    return create_folder(name)


@router.patch("/folders/{folder_id}")
async def watchlist_rename_folder(folder_id: int, body: dict):
    """重命名分组。"""
    name = body.get("name", "")
    if not name:
        raise HTTPException(status_code=422, detail="name is required")
    ok = rename_folder(folder_id, name)
    if not ok:
        raise HTTPException(status_code=404, detail=f"Folder {folder_id} not found")
    return {"success": True}


@router.delete("/folders/{folder_id}")
async def watchlist_delete_folder(folder_id: int):
    """删除分组。"""
    ok = delete_folder(folder_id)
    if not ok:
        raise HTTPException(status_code=404, detail=f"Folder {folder_id} not found")
    return {"success": True}


@router.post("/folders/{folder_id}/stocks")
async def watchlist_add_stock(folder_id: int, body: dict):
    """添加股票到分组。"""
    code = body.get("code", "")
    if not code:
        raise HTTPException(status_code=422, detail="code is required")
    add_stock(folder_id, code)
    return {"success": True}


@router.delete("/folders/{folder_id}/stocks/{code}")
async def watchlist_remove_stock(folder_id: int, code: str):
    """从分组移出股票。"""
    remove_stock(folder_id, code)
    return {"success": True}


@router.get("/folders/{folder_id}/stocks")
async def watchlist_folder_stocks(folder_id: int):
    """分组内股票列表（含行业 + DuckDB 最新收盘价）。"""
    return get_folder_stocks(folder_id)


@router.post("/move")
async def watchlist_move_stock(body: dict):
    """移动股票到其他分组。"""
    from_folder = body.get("from", 0)
    to_folder = body.get("to", 0)
    code = body.get("code", "")
    if not from_folder or not to_folder or not code:
        raise HTTPException(status_code=422, detail="from, to, code are required")
    ok = move_stock(from_folder, to_folder, code)
    if not ok:
        raise HTTPException(
            status_code=404,
            detail=f"Stock {code!r} not found in folder {from_folder}",
        )
    return {"success": True}

File: Front/src/api/modules/watchlist.ts
import client from '../client'
import type { Stock } from '@/api/types'

/** 自选文件夹 */
export interface WatchFolder {
  id: number
  name: string
  codes: string[]
}

/** 获取文件夹列表 */
export function getFolders() {
  return client.get<unknown, WatchFolder[]>('/watchlist/folders')
}

/** 新建文件夹 */
export function createFolder(name: string) {
  return client.post<unknown, WatchFolder>('/watchlist/folders', { name })
}

/** 删除文件夹 */
export function deleteFolder(id: number) {
  return client.delete<unknown, { success: boolean }>(`/watchlist/folders/${id}`)
}

/** 重命名文件夹 */
export function renameFolder(id: number, name: string) {
  return client.patch<unknown, { success: boolean }>(`/watchlist/folders/${id}`, { name })
}

/** 获取某文件夹下的股票（mock 返回全量数组；真实后端将服务端分页） */
export function getFolderStocks(folderId: number) {
  return client.get<unknown, Stock[]>(`/watchlist/folders/${folderId}/stocks`)
}

/** 添加股票到文件夹 */
export function addStock(folderId: number, code: string) {
  return client.post<unknown, { success: boolean }>(`/watchlist/folders/${folderId}/stocks`, {
    code,
  })
}

/** 从文件夹移出股票 */
export function removeStock(folderId: number, code: string) {
  return client.delete<unknown, { success: boolean }>(
    `/watchlist/folders/${folderId}/stocks/${code}`,
  )
}

/** 移动股票到其他文件夹 */
export function moveStock(from: number, to: number, code: string) {
  return client.post<unknown, { success: boolean }>('/watchlist/move', { from, to, code })
}
